File: android/src/main/java/com/reactlibrary/AudioRecorder/AudioRecorderView.java
//
//  AudioRecorderView.java
//  reactnativeaudiorecorder
//
//  Created by Michael Andorfer on 18.09.18.
//  Copyright © 2018 Audvice GmbH. All rights reserved.
//

package com.reactlibrary.AudioRecorder;

import android.content.Context;
import android.graphics.Color;
import android.os.Handler;
import android.widget.RelativeLayout;
import com.reactlibrary.R;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.Timer;
import java.util.TimerTask;

public class AudioRecorderView extends RelativeLayout {
  // The timer which calls the waveform update method
  private Timer timer;
  private boolean isPaused;

  // The task the timer executes
  private TimerTask refreshWaveformTask;

  // The handler of the timer
  final Handler handler = new Handler();

  // The plot which represents the waveform
  private DynamicWaveformView plot;

  // The current amplitude measured by the microphone
  private Float trackedAmplitude;

  // The constructor
  public AudioRecorderView(Context context) {
    super(context);

    // Register the event bus
    EventBus.getDefault().register(this);

    // Apply layout from xml
    inflate(context, R.layout.audio_recorder_view, this);

    // Init the waveform
    this.plot = findViewById(R.id.audio_recorder_waveform);
    this.plot.setStrokeWidth(5);
    this.plot.setAmplitude(0);
    this.plot.setFrequency(0);
    this.plot.setWaveColor();

    // Draw straight line
    resumeWaveform();
  }

  // Method to listen for amplitude changes received from AudioRecordThread
  @Subscribe(threadMode = ThreadMode.MAIN)
  public void onAmplitudeUpdateEvent(AmplitudeUpdateEvent event) {
    this.trackedAmplitude = event.amplitude;
  }

  // Method to listen for wave form events from module
  @Subscribe(threadMode = ThreadMode.MAIN)
  public void onDynamicWaveformEvent(DynamicWaveformEvent event) {
    // Resume the waveform
    if (event.code == 1) {
      isPaused = event.isPaused;
      resumeWaveform();

    } else if (event.code == 2) { // Pause the waveform
      isPaused = event.isPaused;
      pauseWaveform();
      clearWaveform();
    } else if (event.code == 3) { // Set the properties received from RN
      if (event.backgroundColor != null) {
        this.plot.setBackgroundColor(Color.parseColor(event.backgroundColor));
      }

      if (event.lineColor != null) {
        this.plot.setLineColor(Color.parseColor(event.lineColor));
      }
    }
  }

  // Update the waveform according to amplitude change
  private void refreshWaveformWithAmplitude() {
    // Stop further processing if amplitude is not available
    if (this.trackedAmplitude == null) {
      this.trackedAmplitude = 0.0f;
    }

    // Threshold amplitude so that baseline is quite straight
    Float amplitude = (this.trackedAmplitude);
    if(!isPaused) {
      if (amplitude > 8.00f) {
        amplitude = 8.00f;
      } else if (amplitude < 1.50f) {
        amplitude = 0.10f;
      }
    } else {
      amplitude = 0.0f;
    }

    plot.setAmplitude(amplitude);
  }

  // Resume plot, but keep access level private
  public void resumeWaveform() {
    new Thread(new Runnable() {
      public void run() {
      if (timer != null) {
        pauseWaveform();
      }

      // Set number of sinus waves
      plot.setFrequency(4.00f);

      // Create the timer
      timer = new Timer();

      // Initialize the TimerTask's job
      initializeTimerTask();

      // Start the timer for amplitude update processing on waveform
      timer.scheduleAtFixedRate(
          refreshWaveformTask,
          0,
          10
      );
      }
    }).start();
  }

  // Pause plot, but keep access level private
  public void pauseWaveform() {
    if (this.timer != null) {
      this.timer.cancel();
      this.timer = null;
    }
  }

  // Clear plot, but keep access level private
  public void clearWaveform() {
    this.plot.setAmplitude(0);
    this.plot.setFrequency(0);
  }

  // Initialize the waveform (amplitude) update timer task
  public void initializeTimerTask() {
    refreshWaveformTask = new TimerTask() {
      public void run() {
        handler.post(new Runnable() {
          public void run() {
            refreshWaveformWithAmplitude();
          }
        });
      }
    };
  }
}

File: ios/AudioRecorder/AudioRecorderView.swift
//
//  AudioRecorderView.swift
//  reactnativeaudiorecorder
//
//  Created by Michael Andorfer on 24.07.18.
//  Copyright © 2018 Audvice GmbH. All rights reserved.
//

import Foundation
import UIKit

import AudioKit
import AudioKitUI

// Represents the our native ui (view) component
public class AudioRecorderView: EZAudioPlot {
  // The width of the component received from React Native
  public var componentWidth: Double = 0.00
  
  // The height of the component received from React Native
  public var componentHeight: Double = 0.00
  
  // The background color of the view calculated from the received color
  public var bgColor: UIColor = UIColor(white: 1, alpha: 0.0)
  
  // The color of the line
  public var lineColor: UIColor = UIColor(red: 124.0 / 255.0, green: 219.0 / 255.0, blue: 213.0 / 255.0, alpha: 1.0)
  
  // The plot which represents the waveform
  private var plot : SwiftSiriWaveformView = SwiftSiriWaveformView()
  
  // The timer which calls the waveform update method
  private var timer: Timer?
  
  // Tracker which observes the microphone
  private var microphoneTracker = AKMicrophoneTracker()
  
  // Constructor
  private override init(frame: CGRect) {
    // Call super constructor
    super.init(frame: frame)
    
    // Assign frame
    self.frame = frame
    
    // Set width to use 100% (relative)
    self.autoresizingMask = [.flexibleWidth]
  }

  // Initializes our waveform with defined properties
  func setupWaveform(microphoneTracker: AKMicrophoneTracker) {
    DispatchQueue.main.async {
      // Setup plot
      self.setupWaveformPlot(microphoneTracker: microphoneTracker)
    }
  }
  
  // Update the waveform according to amplitude change
  @objc internal func refreshWaveformWithAmplitude(_:Timer) {
    // Simply set the amplitude to whatever you need and the view will update itself.
    DispatchQueue.main.async {
      // Threshold amplitude so that baseline is less sensitive against noise
      // also augment higher amplitude values so that the baseline is more sensitive against speach
      var amplitude = self.microphoneTracker.amplitude * 5
      if (self.microphoneTracker.amplitude < 0.020) {
        amplitude = 0
      }
      
      self.plot.amplitude = CGFloat(amplitude)
    }
  }
  
  // Propagate view update
  public override func layoutSubviews() {
    self.plot.backgroundColor = self.bgColor
    self.plot.waveColor = self.lineColor
  }
  
  // Setup the plot with custom properties
  private func setupWaveformPlot(microphoneTracker: AKMicrophoneTracker) {
    // Set the tracker
    self.microphoneTracker = microphoneTracker
    
    // Create the plot
    self.plot = SwiftSiriWaveformView()
    
    // Set width and height to use 100 % (relative)
    self.plot.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    
    // Set the wave line color
    self.plot.waveColor = self.lineColor
    
    // Set amplitude and frequency to zero to create a straight line
    self.plot.amplitude = 0
    self.plot.frequency = 0
    
    // Define the number of secondary lines
    self.plot.numberOfWaves = 1
    
    // Remove secondary lines
    self.plot.secondaryLineWidth = 0
    
    // Set width of primary line
    self.plot.primaryLineWidth = 5

    // Set the background color of the plot
    self.plot.backgroundColor = self.bgColor
    
    // Set the speed of the wave form
    self.plot.phaseShift = 0.5
    
    // Add the view
    self.addSubview(self.plot)
  }
  
  // Resume plot, but keep access level private
  public func resumeWaveform() {
    // Set number of sinus waves
    self.plot.frequency = 4
    
    DispatchQueue.main.async {
      self.timer = Timer.scheduledTimer(timeInterval: 0.01, target: self, selector: #selector(self.refreshWaveformWithAmplitude(_:)), userInfo: nil, repeats: true)
    }
  }
  
  // Pause plot, but keep access level private
  public func pauseWaveform() {
    DispatchQueue.main.async {
      self.timer?.invalidate()
    }
  }
  
  // Clear plot, but keep access level private
  public func clearWaveform() {
    DispatchQueue.main.async {
      self.plot.amplitude = 0
      self.plot.frequency = 0
    }
  }
  
  required public init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
}
